_includes/utilities/js/helpers/getProp.js

```diff
@@ -1,41 +1,108 @@
 import decodeHtmlEntities from './decodeHtmlEntities.js';
 
-
+// Convert string to either JSON or number, depending on what is passed to 'parse'
```

Comment:
what's this module all about? It looks like super abstract code. It's called getProp and it parses a bunch of stuff and innerHtml? I tried last PR this was touched and now and I still don't understand what this is about.

Reply:
The reason I ask - there's so much code here that's really abstract and I suspect a lot of things rely on how this works - and I wonder if something can't perhpas be done in a more literal, self-explanatory if perhaps slightly more verbose way.

Reply:
Good question. In short this is a shorthand helper that retrieves values from HTML (i.e. for Jekyll) to be passed to components when the are initialised. For example if we have the following:

```
<div data-enhance-component="Example" data-color="green">
  <h3 data-title>Hello world</h3>
  <div data-content>
    <p>This is a paragraph</div>
    <a href="#">This is a link</div>
  </div>
  <button data-button>Button 1</div>
  <button data-button>Button 2</div>
  <button data-button>Button 3</div>
<div>
```

if  `node` is the root `<div>` would just run the following to build me an object to pass to React or JavaScript:
```
const values = {
  color: getProp('color', node, { parse: 'string' }),
  title: getProp('title', node, { parse: 'node', nodeParse: 'string' }),
  content: getProp('content', node, { parse: 'node', nodeParse: 'innerHTML' }),
  buttonsArray: getProp('button', node, { parse: 'node', loop: true }),
}

Reply:
However, you are right that the helper itself is pretty messy at the moment, since I just added use cases to it as it came up. I would like to rewrite to take a graphQL-like approach. I.e.

```
const query = {
  color: 'string'
  title: 'string'
  content: 'innerHTML',
  button: [
     'node'
  ],
}

cosnt value = getProps(node, query);
```

Reply:
right - given your example and 
```
const production = getProps('production', document.body, { parse: 'boolean' });
```
it will look in various places where "production" might be - in attributes or tag contents denoted by empty data attributes, and returns the first? one?

So where the data will come from is ambiguous - I could innocently add an attribute which happens to have the same name.

Doesn't the getProp approach re-impleent what data-attributes are for? perhaps making the markup simpler (less duplication compared to having values in both data- attributes and where you want them for real in the markup) or the javascript simpler (e.g. not pulling values from the text contents of elements) but risking surprising results?

Reply:
oh wait - production is also `data-production` in the markup - so based on arguments it will get data from the contents or the data-attributes?

Reply:
So do I understand correctly that getProp is about initialising react from the markup? How do other people do it?

Reply:
So the above example will see if there is a `data-production`  exclusively on the `document.body` tag. It's only if you pass `{ parse: 'node` }` that it will look for an inner-node inside the node that you passed. (As mentioned the API is not the cleanest at the moment).

I won't say it re-implements `getAttribute`, it's merely a wrapper around `getAttribute` and a lot of other boilerplate tasks happening before and after `getAttribute` (such as find DOM elements, looping through DOM elements, and parsing values).

This saves me about 30-40 lines of reduce/map functions and lamdas when I pass info from Jekyll to JS.

Reply:
That makes sense. Is this referring to the react concept _prop_?

I think if you document that the module is about getting data from the markup (which happens to be produced using jekyll/by hand) to React to provide Props for initialising the components, it'd really help people like me to be able to make sense of what's going on here. Then I've got something to google and quickly understand it.

It looks like (three years ago at least) you'd be on your own and have to implement stuff like this, but we're introducing a "system" which is worth making clear at the top of the "module" or whatever the file is called. https://www.reddit.com/r/reactjs/comments/2xckya/initialize_a_react_component_with_prerendered_html/

Additionally, because this is abstract code the name can't really encompass what it does, what the inputs and outputs are, so it's worth documenting them clearly. I wouldn't document every function but I needed a complete lecture on these :D Thanks for the education

Reply:
> That makes sense. Is this referring to the react concept prop?

Jip, refers to React concept of _prop_. 👍 

> So do I understand correctly that getProp is about initialising react from the markup? How do other people do it?

The short answer is most people don't 😜.  More often NodeJS builds the initial state server-side or in the build process (like we're doing with Public People), and then it just gets hydrated in the client-side (https://medium.com/@baphemot/understanding-reactjs-data-hydration-initialization-bacbb790c7cb).

However, there are cases where you're using React to 'enhance' markup generated from another language/templating system. I know that Wordpress is incrementally replacing parts of their PHP stack with React, so I'm sure they probably also have some in-house way of connecting JS to PHP.

My approach with using data-attributes as hooks for React is very much inspired by the work of Lea Verou (a sitting member W3C CSS Working Group): https://www.smashingmagazine.com/2017/02/designing-html-apis/. Some people use CSS classes or ID's as hooks for JS, but that can very often turn into a nightmare when refactoring HTML if you don't know what affects CSS and what affects behaviour. I prefer reserving classes exclusively for controlling styling and data-attributes for controlling behaviour.

Reply:
Cool. Initially I was afraid this could get data from data-attributes, the contents, or just the _existence_ of normal attributes (boolean). I understand now that the latter isn't the case.

I agree with your approach now I know what's going on.